src/control/cmd/daos_server/storage.go

```diff
@@ -36,95 +36,179 @@ type storagePrepareCmd struct {
+			msg = "PMem namespaces removed and AppDirect interleaved regions will be removed. "
 		}
+		cmd.log.Info(msg + storage.ScmMsgRebootRequired)
 
-		// Prepare NVMe access through SPDK
-		if _, err := cmd.scs.NvmePrepare(storage.BdevPrepareRequest{
-			HugePageCount: cmd.NrHugepages,
-			TargetUser:    cmd.TargetUser,
-			PCIAllowList:  cmd.PCIAllowList,
-			PCIBlockList:  cmd.PCIBlockList,
-			Reset_:        cmd.Reset,
-			EnableVMD:     true, // vmd will be prepared if available
-		}); err != nil {
-			scanErrors = append(scanErrors, err)
-		}
+		return nil
 	}
 
-	scmScan, err := cmd.scs.ScmScan(storage.ScmScanRequest{})
-	if err != nil {
-		return common.ConcatErrors(scanErrors, err)
+	if state == storage.ScmStateUnknown {
+		return scanErrors.add(errors.New("failed to report state"))
+	}
+	if state == storage.ScmStateNoModules {
+		return scanErrors.add(storage.FaultScmNoModules)
 	}
 
-	if prepScm && len(scmScan.Modules) > 0 {
-		cmd.log.Info(op + " locally-attached SCM...")
-
-		if err := cmd.CheckWarn(cmd.log, scmScan.State); err != nil {
-			return common.ConcatErrors(scanErrors, err)
+	if cmd.Reset {
+		// Respond to resultant state on prepare reset.
+		if state == storage.ScmStateNoRegions {
+			cmd.log.Info("SCM reset successfully!")
+			return nil
 		}
 
-		// Prepare SCM modules to be presented as pmem device files.
-		// Pass evaluated state to avoid running GetScmState() twice.
-		resp, err := cmd.scs.ScmPrepare(storage.ScmPrepareRequest{Reset: cmd.Reset})
-		if err != nil {
-			return common.ConcatErrors(scanErrors, err)
+		return scanErrors.add(errors.Errorf("unexpected state %q after prepare reset", state))
+	}
+
+	// Respond to resultant state on prepare setup.
+	switch state {
+	case storage.ScmStateNoRegions:
+		scanErrors.add(errors.New("failed to create regions"))
+	case storage.ScmStateNotInterleaved:
+		scanErrors.add(storage.FaultScmNotInterleaved)
+	case storage.ScmStateFreeCapacity:
+		scanErrors.add(errors.New("failed to create namespaces"))
+	case storage.ScmStateNoFreeCapacity:
+		if len(resp.Namespaces) == 0 {
+			scanErrors.add(errors.New("failed to find namespaces"))
+			break
 		}
-		if resp.RebootRequired {
-			cmd.log.Info(storage.ScmMsgRebootRequired)
-		} else if len(resp.Namespaces) > 0 {
-			var bld strings.Builder
-			if err := pretty.PrintScmNamespaces(resp.Namespaces, &bld); err != nil {
-				return err
-			}
-			cmd.log.Infof("SCM namespaces:\n%s\n", bld.String())
-		} else {
-			cmd.log.Info("no SCM namespaces")
+		// Namespaces created successfully.
+		var bld strings.Builder
+		if err := pretty.PrintScmNamespaces(resp.Namespaces, &bld); err != nil {
+			return err
 		}
-	} else if prepScm {
-		cmd.log.Info("No SCM modules detected; skipping operation")
+		cmd.log.Infof("SCM namespaces:\n%s\n", bld.String())
+	default:
+		scanErrors.add(errors.Errorf("unexpected state %q after prepare reset", state))
 	}
 
-	if len(scanErrors) > 0 {
-		return common.ConcatErrors(scanErrors, nil)
+	return nil
```

Comment:
I don't like scanErrors usage either particularly and I'm going to remove it in a subsequent PR when we separate `storage prepare` into `storage (prep-scm|prep-nvme)`.
The reason it exists is in the case where we are doing both prepare for SCM and NVMe in the same command (no --*-only option specified), we want to perform both operations regardless of a failure in either. Therefore we don't return immediately on error and just store in scanErrors which gets processed at the end.
Wrt return values, the rationale is that for an error that is specific to the storage operation e.g. NVMe operation fails, the error should be added to scanErrors but should not prevent the subsequent prepScm call so nil is returned from prepNvme. If on the other hand the error is not a result of storage operation but is a system unexpected problem e.g. current user cannot be retrieved, then we want to return error from prepNvme and fail out before prepScm gets called.
Anyway this ugliness will go away soon.

Reply:
The suggestion of "have the caller aggregate them if needed" may not be substantially simpler because you would still need to work out which error should prevent proceeding onto the next step, i.e. differentiate between system and storage errors requiring some sort of lookup.

Reply:
Could create a type scanError for all the scan-specific errors and use `errors.Is` to check it, if we have these two separate categories.If it's going away soon it doesn't matter that much, though.

Reply:
Will leave as it will be removed in follow-up.